bin/propolis-server/src/lib/stats/virtual_disk.rs

```diff
@@ -0,0 +1,381 @@
+    }
+
+    fn on_flush_completion(
+        &mut self,
+        result: block::Result,
+        duration: Duration,
+    ) {
+        let index = match result {
+            block::Result::Success => {
+                let _ = self.io_latency[FLUSH_INDEX]
+                    .datum
+                    .sample(duration.as_nanos() as u64);
+                self.flushes.datum += 1;
+                return;
+            }
+            block::Result::Failure => FAILURE_INDEX,
+            block::Result::ReadOnly => READONLY_INDEX,
+            block::Result::Unsupported => UNSUPPORTED_INDEX,
+        };
+        self.failed_flushes[index].datum.increment();
+    }
+}
+
+/// Number of I/O kinds we track.
+const N_IO_KINDS: usize = 3;
+
+/// Indices into arrays tracking operations broken out by I/O kind.
+const READ_INDEX: usize = 0;
+const WRITE_INDEX: usize = 1;
+const FLUSH_INDEX: usize = 2;
+
+/// String representations of I/O kinds we report to Oximeter.
+const READ_KIND: &str = "read";
+const WRITE_KIND: &str = "write";
+const FLUSH_KIND: &str = "flush";
+
+/// Number of failure kinds we track.
+const N_FAILURE_KINDS: usize = 3;
+
+/// Indices into arrays tracking operations broken out by failure kind.
+const FAILURE_INDEX: usize = 0;
+const READONLY_INDEX: usize = 1;
+const UNSUPPORTED_INDEX: usize = 2;
+
+/// String representations of failure kinds we report to Oximeter.
+const FAILURE_KIND: &str = "failed";
+const READONLY_KIND: &str = "read-only";
+const UNSUPPORTED_KIND: &str = "unsupported";
+
+/// Latency is measured in nanoseconds. We want to track between 1 microsecond,
+/// which is 10 ** 3 nanos, and 10s, which is 10 * 1e9 == 10 ** 10 nanoseconds.
+const LATENCY_POWERS: (u16, u16) = (3, 10);
+
+/// Sizes are measured in powers of 2 from 4KiB to 1GiB
+const SIZE_POWERS: (u16, u16) = (12, 30);
+
+/// A [`Producer`] that emits statistics about virtual disks.
+///
+/// This type is shared between the block devie that handles guest I/Os, and the
+/// oximeter producer server, which publishes data to oximeter when it polls us.
+/// As I/Os are completed, the [`VirtualDiskProducer::on_completion()`] method
+/// is called, to update stats with the results of the I/O operation.
+///
+/// As oximeter polls us, the producer server also collects these updated
+/// statistics.
+#[derive(Clone, Debug)]
+pub struct VirtualDiskProducer {
+    // Shareable inner type actually managing the stats.
+    inner: Arc<Mutex<VirtualDiskStats>>,
+}
+
+impl VirtualDiskProducer {
+    /// Create a producer to track a virtual disk.
+    pub fn new(
+        block_size: u32,
+        instance_id: Uuid,
+        disk_id: Uuid,
+        metadata: &InstanceMetadata,
+    ) -> Self {
+        let disk = VirtualDisk {
+            attached_instance_id: instance_id,
+            block_size,
+            disk_id,
+            project_id: metadata.project_id,
+            silo_id: metadata.silo_id,
+            sled_id: metadata.sled_id,
```

Comment:
The `sled_id` here is going to be the sled the disk was attached to when it was in use, right?
We don't want to confuse that with the location(s) of all the downstairs that could be what underpins the virtual disk that Propolis sees.

Reply:
That's right, it's the ID of the sled hosting the instance. I think the description in the `virtual-disk.toml` schema should make this clear, but let me know if that's lacking in some way.